src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -1053,12 +1003,14 @@ SDCPUThread::predictCtrlInst(shared_ptr<InflightInst> inst_ptr)
 {
     DPRINTF(SDCPUBranchPred, "Requesting branch predictor for control\n");
 
-    const InstSeqNum seqnum = inst_ptr->seqNum();
-    inst_ptr->addSquashCallback([this, seqnum] {
-        _cpuPtr->getBranchPredictor()->squash(seqnum, threadId());
+    const weak_ptr<InflightInst> weak_inst = inst_ptr;
+    inst_ptr->addSquashCallback([this, weak_inst] {
```

Comment:
This is redundant, because this callback belongs to inst_ptr. weak_inst will never fail to lock, otherwise we would be unable to call the squash callback. We didn't capture inst_ptr either because we didn't need a reference to it to call squash, as long as we had the sequence number...

Reply:
It's a more lightweight capture to just capture an unmanaged reference to a SDCPUThread and the seqnum than a managed weak_ptr.

Reply:
and the callback will already naturally be cleaned up or not called, since it's purely managed by the inst_ptr's callback list.